File: 2015/Day7/main.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <unordered_map>
#include <vector>
#include <algorithm>

using namespace std;

unordered_map<string, uint16_t> wire_values;
unordered_map<string, vector<string>> instructions;

uint16_t get_wire_value(const string &wire)
{
    if (wire_values.find(wire) != wire_values.end())
    {
        return wire_values[wire];
    }
    if (isdigit(wire[0]))
    {
        return stoi(wire);
    }

    vector<string> &instruction = instructions[wire];
    uint16_t result;

    if (instruction.size() == 1)
    {
        result = get_wire_value(instruction[0]);
    }
    else if (instruction.size() == 2)
    {
        result = ~get_wire_value(instruction[1]);
    }
    else
    {
        uint16_t left = get_wire_value(instruction[0]);
        uint16_t right = get_wire_value(instruction[2]);
        string op = instruction[1];

        if (op == "AND")
            result = left & right;
        else if (op == "OR")
            result = left | right;
        else if (op == "LSHIFT")
            result = left << right;
        else if (op == "RSHIFT")
            result = left >> right;
    }

    wire_values[wire] = result;
    return result;
}

uint16_t solve_circuit(bool override_b = false, uint16_t b_value = 0)
{
    wire_values.clear();
    if (override_b)
    {
        wire_values["b"] = b_value;
    }
    return get_wire_value("a");
}

int main()
{
    ifstream file("input.txt");
    string line;

    while (getline(file, line))
    {
        istringstream iss(line);
        vector<string> tokens;
        string token;

        while (iss >> token)
        {
            if (token != "->")
            {
                tokens.push_back(token);
            }
        }

        string wire = tokens.back();
        tokens.pop_back();
        instructions[wire] = tokens;
    }

    uint16_t part_1 = get_wire_value("a");
    cout << "Part 1: " << part_1 << endl;
    uint16_t part_2 = solve_circuit(true, part_1);
    cout << "Part 2: " << part_2 << endl;

    return 0;
}

File: 2015/Day7/main.py
def parse_instruction(instruction):
    parts = instruction.split(' -> ')
    output = parts[1].strip()
    input_parts = parts[0].split()

    if len(input_parts) == 1:
        return output, ('ASSIGN', input_parts[0])
    elif len(input_parts) == 2:
        return output, ('NOT', input_parts[1])
    else:
        return output, (input_parts[1], input_parts[0], input_parts[2])

def evaluate(wire, circuit, memo):
    if (wire.isdigit()):
        return int(wire)
    if wire in memo:
        return memo[wire]
    
    operation = circuit[wire]

    if operation[0] == 'ASSIGN':
        result = evaluate(operation[1], circuit, memo)
    elif operation[0] == 'NOT':
        result = ~evaluate(operation[1], circuit, memo) & 0xffff
    elif operation[0] == 'AND':
        result = evaluate(operation[1], circuit, memo) & evaluate(operation[2], circuit, memo)
    elif operation[0] == 'OR':
        result = evaluate(operation[1], circuit, memo) | evaluate(operation[2], circuit, memo)
    elif operation[0] == 'LSHIFT':
        result = evaluate(operation[1], circuit, memo) << evaluate(operation[2], circuit, memo) & 0xffff
    elif operation[0] == 'RSHIFT':
        result = evaluate(operation[1], circuit, memo) >> int(operation[2])
    
    memo[wire] = result
    return result

def solve_circuit(instructions, override_b=None):
    circuit = {}
    for instruction in instructions:
        output, operation = parse_instruction(instruction)
        circuit[output] = operation

    if override_b is not None:
        circuit['b'] = ('ASSIGN', str(override_b))

    memo = {}
    return evaluate('a', circuit, memo)

with open('input.txt', 'r') as file:
    instructions = file.read().splitlines()

part1 = solve_circuit(instructions)
print(f"Part 1: {part1}")
print(f"Part 2: {solve_circuit(instructions, override_b=part1)}")